### PR TITLE
Add failure action for rolling updates

### PR DESCRIPTION
**Pull request by aaronlehmann**
This changes the default behavior so that rolling updates will not proceed once an updated task fails to start, or stops running during the update. Users can use `docker service inspect --pretty servicename` (btw, why `--pretty`? I think we should change this) to see the update status, and if it pauses due to a failure, it will explain that the update is paused, and show the task ID that caused it to pause. It also shows the time since the update started.

A new `--update-on-failure=(pause|continue)` flag selects the behavior. Pause means the update stops once a task fails, continue means the old behavior of continuing the update anyway.

In the future this will be extended with additional behaviors like automatic rollback, and flags controlling parameters like how many tasks need to fail for the update to stop proceeding. This is a minimal solution for 1.12.

ping @cpuguy83 @aluzzardi @stevvooe @tiborvass @sfsmithcha

### CODE REVIEW

**Comment by aaronlehmann** [comment]
Example output in `docker service inspect --pretty`:

```
Update status:
 State:	paused
 Started:	2 seconds ago
 Message:	update paused due to failure or early termination of task 11rq7a5e65cnnmwlbs3jaajeg
```

**Comment by aaronlehmann** [comment]
swarmkit PR was merged. swarmkit is being properly vendored now.

engine-api PR is docker/engine-api#333

PTAL

**Comment by aaronlehmann** [comment]
This should be ready to go. PTAL.

File: contrib/completion/zsh/_docker [hunk 1]
```diff
@@ -1094,6 +1094,7 @@ __docker_service_subcommand() {
         "($help)--restart-window=[Window used to evaluate the restart policy]:window: "
         "($help)--stop-grace-period=[Time to wait before force killing a container]:grace period: "
         "($help)--update-delay=[Delay between updates]:delay: "
+	"($help)--update-on-failure=[Action on update failure]:mode:(pause continue)"
```
Comment (aluzzardi) [inline contrib/completion/zsh/_docker]: nit: Indent

**Comment by aluzzardi** [comment]
LGTM

We should consider doing something in `docker service ls`, but I believe that should be in a different PR. Proposal:

```
ID            NAME  REPLICAS  IMAGE                  UPDATE STATUS              COMMAND
55aiqua4y8wi  vote  10/10     instavote/vote:indent  Updating 18 minutes ago
55aiqua4y8wi  vote  10/10     instavote/vote:indent  Completed 2 hours ago
```

**Comment by abronan** [comment]
LGTM

**Comment by thaJeztah** [comment]
@aaronlehmann does this help with https://github.com/docker/docker/issues/24092 as well?

File: api/client/service/opts.go [hunk 2]
```diff
@@ -545,6 +548,7 @@ const (
 	flagRestartWindow      = "restart-window"
 	flagStopGracePeriod    = "stop-grace-period"
 	flagUpdateDelay        = "update-delay"
+	flagUpdateOnFailure    = "update-on-failure"
```
Comment (cpuguy83) [inline api/client/service/opts.go]: As I mentioned in swarmkit, I feel like this flag name is really really ambiguous.
`update-failure-action`, `on-update-failure`, `update-failure-policy`

**Comment by cpuguy83** [comment]
Looks like failure action is a required argument as the API does not assume a default. Is this intentional?

File: api/client/service/opts.go (hunk 2 above)
Comment (aaronlehmann) [inline api/client/service/opts.go]: It was originally `update-failure-action` in swarmkit. I changed it based on this comment: https://github.com/docker/swarmkit/pull/1204#discussion_r71886182 :)

I'll change it back to `update-failure-action`.

**Comment by aaronlehmann** [comment]
> Looks like failure action is a required argument as the API does not assume a default. Is this intentional?

If a value is omitted in the API, the protobuf field will end up with the 0 value, which is UpdateConfig_PAUSE.

**Comment by cpuguy83** [comment]
@aaronlehmann I got an error response saying I gave an invalid configuration when I omitted it.

**Comment by aaronlehmann** [comment]
Oh, you're right, there's a default case that returns an error.

I'll change this to handle the missing parameter gracefully.

```go
 +		default:
 +			return swarmapi.ServiceSpec{}, fmt.Errorf("unrecongized update failure action %s", s.UpdateConfig.FailureAction)
```

**Comment by aaronlehmann** [comment]
@cpuguy83: Addressed comments. PTAL.

**Comment by tiborvass** [comment]
LGTM

**Comment by aaronlehmann** [comment]
@thaJeztah: It is not related to that issue.

**Comment by cpuguy83** [comment]
LGTM

**Comment by cpuguy83** [comment]
ping @thaJeztah for docs

**Comment by thaJeztah** [comment]
docs changes in this PR look good, but /cc @sfsmithcha

**Comment by sfsmithcha** [comment]
docs changes LGTM

**Pull request merged**